Allow use of Pre-Release Extensions

mark-lunarg opened the pull request:
This set of changes will allow unique-object to handle non-vulkan extension stuctures as part of a pnext chain.  It will NOT suppress/prevent warnings for using unrecognized extensions, and it does NOT allow for pre-release extension structs that contain Vulkan objects.

@tobine, please take a look and let me know if this sufficiently addresses your issue.

Bonus:  As a side-effect, `vk_safe_structs.cpp` is no longer in-lined in other source modules.

Fixes #1820.

tobine commented on layers/generated/chassis.cpp:
```diff
@@ -286,31 +286,36 @@ void SetValidationFeatures(CHECK_DISABLED &disable_data, CHECK_ENABLED &enable_d
     }
 }
 
+std::string GetNextToken(std::string &token_list, std::string &delimiter, size_t &pos) {
```
Recommend "token_list" is ptr rather than non-const ref.
delimiter and pos can be const ref.

sugoi1 replied:
pos is written to, so it can't be const

tobine commented on scripts/layer_chassis_generator.py:
```diff
@@ -904,31 +904,36 @@ class ValidationObject {
     }
 }
 
+std::string GetNextToken(std::string &token_list, std::string &delimiter, size_t &pos) {
```
Same comment as above, Recommend "token_list" is ptr rather than non-const ref.
delimiter and pos can be const ref.

sugoi1 replied:
pos is also written to in this function, so it can't be const

tobine commented on BUILD.gn:
```diff
@@ -83,8 +83,6 @@ config("vulkan_layer_config") {
 
 core_validation_sources = [
-  # This file is manually included in the layer
-  # "layers/generated/vk_safe_struct.cpp",
```
That's an awesome bonus feature!

tobine commented on layers/generated/chassis.cpp:
```diff
@@ -359,24 +362,55 @@ void CreateFilterMessageIdList(std::string raw_id_list, std::string delimiter, s
+    }
+}
+
+uint32_t TokenToUint(std::string &token) {
```
token can be const ref.

mark-lunarg commented:
Will fix the pointers/const refs.  The list can be set through the settings file, or an environment variable.  If that won't work, we'll need to add support for Android properties (I had assumed this was just for desktop/testing). 

For the pNext sizes, if the appropriate structure size is used there's no issue -- this might mean having platform-specific lists. Another options is to  try to document that the size is without the pnext (which is not so  great) -- what if the struct contains other void pointers?  Ultimately, we really need to know the correct size.  

As previously mentioned, perhaps the best route is to create an EXT extension that would include metadata explicitly describing the new structure, including any objects that need to be wrapped.

tobine commented:
Thinking a bit about the EXT suggested.
This would include something like an "ALLOWED_STRUCTS" ext struct that includes unique struct id and size for any structs that validation layers should allow through? And some type of additional, optional data to indicate objects in the struct that have to be wrapped?
Then this would be passed in at instance create time and stored by the validation layers so that they correctly handle the specified structs?

If I'm thinking of it correctly then that seems like a good solution as the app itself will be passing down the info and will have the correct size for any structs that it uses given the underlying architecture.

mark-lunarg commented:
@tobine, rebased, added commit to allow reads from the registry, so you might give this one a shot.  Will continue to pursue a layer-only extension for setting any layer property programmatically.

The arguments to the setting are a pair of unsigned integers designating the stype value of the custom struct and the size in bytes of that stuct.  So to declare two pre-release extensions on Android:

`adb shell setprop debug.khronos_layer.custom_stype_list 30030408,16,0x20001011,32`

On desktop, use the settings file or env var:
`khronos_layer.custom_stype_list=30030408,16,0x20001011,32`
`set VK_LAYER_CUSTOM_STYPE_LIST=30030408,16,0x20001011,32`

amaiorano commented:
Just to let you know, I've been testing this change, both setting env. var and with the vk_layer_settings file, and it seems to be working fine. Once the `return output.c_str()` problem is fixed, I'll test again and make sure it passes our internal build machines, and if all goes well, we could merge to master. Thanks!

mark-lunarg commented:
@amaiorano, @tobine, thanks for the review!  We ran into some problems here with the 'setenv' call on Android, and were planning to redo portions of this change on top of the private extension PR, #1920.  This is just to get testing on Android unblocked on our end, but is preferable (for us, at least) to the setenv-based stuff anyway. 

That refactor shouldn mainly affect the test and the fetch part of the settings, but we'll make sure the vk_layer_config issues are addressed at that time.

amaiorano commented:
Does this mean we will still be able to use the env. var. for setting custom extensions as per this PR?

Also, do you have a rough ETA for when this change will be rebased/updated?

Cheers!

mark-lunarg commented:
Yes, that'll all still work on desktop environments.  Just curious, did it work for you on Android?  It should be this week, working on getting the settings ext in today maybe.

amaiorano commented:
I didn't test at all on Android. Unfortunately, I'm not setup for that. Also, for our purposes, we presently need this feature for non-Android targets. Having said that, our build bots compile and run tests on Android, and that's where the return stack var warning came from.

mark-lunarg commented:
@tobine, @amaiorano, rebased on top of the private settings extension.  _Should_ work the same as before, but now we have tests for this feature that work on Android.  This no longer implements the use of Android properties, but I can and may add that at a future time and will be sure to fix the warnings beforehand.

amaiorano commented:
Thank you, @mark-lunarg, I'll test against this updated commit today.

mark-lunarg commented:
Passed our internal CI and we have a green board here.  Once you sign off I'll push it.  Thanks!

amaiorano commented:
Hi Mark,

Ran into a problem on x86 release builds of our angle_end2end_tests when using this change. Am getting some memory corruption during `loader_create_device_chain`:

```
 	ntdll.dll!_RtlReportCriticalFailure@12()	Unknown
 	ntdll.dll!_RtlpReportHeapFailure@4()	Unknown
 	ntdll.dll!_RtlpHpHeapHandleError@12()	Unknown
 	ntdll.dll!_RtlpLogHeapFailure@24()	Unknown
 	ntdll.dll!_RtlpFreeHeapInternal@20()	Unknown
 	ntdll.dll!RtlFreeHeap()	Unknown
 	ucrtbase.dll!75c5e58b()	Unknown
 	ucrtbase.dll![Frames below may be incorrect and/or missing, no symbols loaded for ucrtbase.dll]	Unknown
 	ucrtbase.dll!75c5e558()	Unknown
 	nvoglv32.dll!059563ab()	Unknown
 	vulkan-1.dll!loader_create_device_chain() Line 6175	C
 	vulkan-1.dll!loader_layer_create_device() Line 5783	C
 	vulkan-1.dll!vkCreateDevice() Line 779	C
 	libGLESv2.dll!rx::RendererVk::initializeDevice() Line 1285	C++
 	vulkan-1.dll!vkGetPhysicalDeviceQueueFamilyProperties() Line 766	C
```

I can't reproduce this on our x64 build, and for x86, it only happens when I compile in Release. If I enable our Debug build, the problem does not occur. I will try to get more useful information tomorrow, but I wanted to give you a heads up in case you had an idea of what's happening.

mark-lunarg commented:
Thanks!  Will write a test and look at it in the morning.

amaiorano commented:
Since the crash is happening in the loader, I'm wondering if maybe there's an incompatibility between the work on this branch, and the version of the loader we're using. We are currently at commit `24a17f5fe4efc1fe4338189959d72433ecafe484` of vulkan-loader.

mark-lunarg commented:
@amaiorano, wrote a test this morning, built 32-bit release versions of layers and loader, but am unable to reproduce any failures with the custom structure specified. 
One thing to note is that the desktop loader prepends the pNext chain with a couple of secret loader structs (see `VK_STRUCTURE_TYPE_LOADER_DEVICE_CREATE_INFO `for example) though this should be benign.  
I also don't have `libGLESv2.dll` in my call stack.  Do you have any other suggestions? Does it happen if you run `relwithdebinfo`?

amaiorano commented:
Hi Mark,

I did some more debugging, and have some useful information. The crash occurs because at one point, `custom_stype_info` ends up with duplicates. When this happens, the loop at the bottom of `FreePnextChain` that frees custom types ends up freeing the same node more than once.

I'm not entirely sure I understand the lifetime management of the global `custom_stype_info` vector, but I do see that the code calls push_back on it, but never actually clears it or removes elements from it. If this is expected, then I suppose a fix would be to ensure duplicates are never added to the list.

It's easy enough to repro this bug by adding the same id,size pair twice in the env var or settings file (or the same pair in each). However, that's not what's happening in my case. Let me know if you want me to keep debugging this to understand how we end up with dupes in the list.

mark-lunarg commented:
@amaiorano, thank you for this information, it is super-useful.  I will add a workaround to prevent duplicate custom sTypes.  I'm also extremely curious to see where the root cause lies, it could very well be in the config or option handling somewhere.

mark-lunarg commented:
@amaiorano, I've added the duplicate-prevention checks and changed one of the tests to include a duplicate.  Please let me know how it's looking, Thanks!

amaiorano commented:
Thanks, @mark-lunarg, going to try it right now.

amaiorano commented:
Our previous failed tests have passed, so it looks like it's working now. Am running all our presubmit tests again to make sure they all pass. Will let you know.

amaiorano commented:
Hi @mark-lunarg, all presubmit tests are passing now! You can go ahead and merge this to master. Thanks!

mark-lunarg commented:
Rebased and pushed to Master.  Thanks a lot for all the help, @amaiorano!
